modules/samples/virtual_firealarm/src/org.wso2.carbon.device.mgt.iot.sample.virtual.firealarm.service.impl/src/main/java/org/wso2/carbon/device/mgt/iot/sample/virtual/firealarm/service/impl/util/mqtt/MQTTClient.java

```diff
@@ -0,0 +1,280 @@
+					"\n\tException: " + ex;
+			if (log.isDebugEnabled()) {
+				log.debug(errorMsg);
+			}
+			throw new DeviceManagementException(errorMsg, ex);
+		}
+	}
+
+	/**
+	 * Callback method which is triggered once the MQTT client losers its connection to the broker.
+	 * A scheduler thread is spawned to continuously re-attempt and connect to the broker and
+	 * subscribe to the device's topic. This thread is scheduled to execute after every break
+	 * equal to that of the 'reConnectionInterval' of the MQTTClient.
+	 *
+	 * @param throwable a Throwable Object containing the details as to why the failure occurred.
+	 */
+	public void connectionLost(Throwable throwable) {
+		log.warn("Lost Connection for client: " + this.clientId + " to " + this.mqttBrokerEndPoint + ".\nThis was due to - " + throwable.getMessage());
+
+		Runnable reSubscriber = new Runnable() {
+			@Override
+			public void run() {
+				if (!isConnected()) {
+					if (log.isDebugEnabled()) {
+						log.debug("Subscriber reconnecting to queue........");
+					}
+					try {
+						connectAndSubscribe();
+					} catch (DeviceManagementException e) {
+						if (log.isDebugEnabled()) {
+							log.debug("Could not reconnect and subscribe to ControlQueue.");
+						}
+					}
+				} else {
+					return;
+				}
+			}
+		};
+
+		ScheduledExecutorService service = Executors.newSingleThreadScheduledExecutor();
+		service.scheduleAtFixedRate(reSubscriber, 0, this.reConnectionInterval, TimeUnit.SECONDS);
+	}
+
+	/**
+	 * Callback method which is triggered upon receiving a MQTT Message from the broker. Spawns a
+	 * new thread that executes any actions to be taken with the received message.
+	 *
+	 * @param topic       the MQTT-Topic to which the received message was published to and the
+	 *                    client was subscribed to.
+	 * @param mqttMessage the actual MQTT-Message that was received from the broker.
+	 */
+	public void messageArrived(final String topic, final MqttMessage mqttMessage) {
+		Thread subscriberThread = new Thread() {
+			public void run() {
+				postMessageArrived(topic, mqttMessage);
+			}
+		};
+		subscriberThread.start();
+	}
+
+	/**
+	 * Callback method which gets triggered upon successful completion of a message delivery to
+	 * the broker.
+	 *
+	 * @param iMqttDeliveryToken the MQTT-DeliveryToken which includes the details about the
+	 *                           specific message delivery.
+	 */
+	public void deliveryComplete(IMqttDeliveryToken iMqttDeliveryToken) {
+		String message = "";
+		try {
+			message = iMqttDeliveryToken.getMessage().toString();
+		} catch (MqttException e) {
+			log.error("Error occurred whilst trying to read the message from the MQTT  delivery token.");
+		}
+		String topic = iMqttDeliveryToken.getTopics()[0];
+		String client = iMqttDeliveryToken.getClient().getClientId();
+		log.info("Message - '" + message + "' of client [" + client + "] for the topic (" + topic + ") was delivered successfully.");
```

Comment:
If an MqttException is thrown, message variable will not be set. When that is the case, referring to that later will not be fruitful which you actually do in line 257. Please rethink on the logic here.